PostgreSQL@12: Use its own data dir and log files when run via services

cypher opened the pull request:
This updates the plist definition for `brew services` to work analogous to the one for PostgreSQL@11, so it uses `/usr/local/var/postgresql@12` instead of `/usr/local/var/postgres`, which collides with the existing PostgreSQL 13 formula.

(Note: This does not change anything about the formula itself, only the plist definition)

-----

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

SMillerDev commented on Formula/postgresql@12.rb:
```diff
@@ -115,16 +115,16 @@ def plist
           <string>#{opt_bin}/postgres</string>
           <string>-D</string>
-          <string>#{var}/postgres</string>
+          <string>#{var}/#{name}</string>
```
These directories also need to be created in `post_install` then, and updated in the caveat.

SMillerDev reviewed (commented):
Once this one is complete, could you also backport this to other postgres versions?

cypher commented:
@SMillerDev I've updated the PR with the requested changes, thanks!

I'll also check the other formulaes, and create new PRs as needed.

MikeMcQuaid reviewed (commented):
I'm fine with this but it looks to me like this is going to break existing installs, no?

cypher commented:
That is true. I've been manually tweaking the formula every time it changes because I need multiple versions running side-by-side at the same time, and it's the only one of the Postgres-formulaes that has this issue. I have no idea how many other people are only running this PosgreSQL-formula.

A possible fix would be to examine the default data dir, look inside the PG_VERSION file, and if it contains the string "12", to either emit a warning message, or use the old data dir path in the plist file.

MikeMcQuaid commented:
The way we've done stuff like this in the past is, yeh, to default to the old data directory if it exists, printing a caveat/warning if necessary and using the new/versioned one if not. See https://github.com/Homebrew/homebrew-core/pull/60190 for something related.

cypher commented:
So I've updated the formula to print different kinds of warnings in `caveats`:

* If the old data dir is detected to be using version 12 (ie., `/usr/local/var/postgres/PG_VERSION` contains the string `12`), and the `postgresql` formula is not installed, it prints a warning about possible conflicts, and how to migrate the data dir to the versioned path
* If the old data dir is detected to be using version 12, and the `postgresql` formula is also installed, it also prints a warning, but no migration instructions, since I wasn't sure if it was still save to just move the data dir in this case.
* If the old data dir is detected, but PG_VERSION doesn't contain `12`, then no warning is printed, because presumably the user has solved the conflicting data path problem (or isn't actually using the `postgresql@12` formula).

The plist method also makes sure to use the versioned data dir, if it exists, falling back to the old one. Same with the log files.

So basically, this should display a warning if anyone is using this formula, and also might have installed the postgresql formula.

MikeMcQuaid commented on Formula/postgresql@12.rb:
```diff
@@ -81,22 +81,87 @@ def post_install
+  # postgresql formula. So we check whether the versioned data dir exists
+  # and has a PG_VERSION file, which should indicate that the versioned
+  # data dir is in use. Otherwise, returns the old data dir path.
+  def postgresql_datadir
```
This should be used in `post_install`, I think?

cypher replied:
So I didn't use this method, but I added a new helper method that always resolves to the versioned data dir path, based on my understanding that `post_install` is only used when the formula is installed. This way, I hope it's clear what the intended data dir path is supposed to be when someone is reading the formula source.

Addressed in 03dfafe1a4

MikeMcQuaid commented on Formula/postgresql@12.rb:
```diff
@@ -81,22 +81,87 @@ def post_install
+  end
+
+  def versioned_data_dir_exists?
+    File.exist?("#{var}/#{name}/PG_VERSION")
```
This should be used in `post_install`, I think?

cypher replied:
Addressed in 03dfafe1a4

MikeMcQuaid commented on Formula/postgresql@12.rb:
```diff
@@ -81,22 +81,87 @@ def post_install
+    if versioned_data_dir_exists?
+      "#{var}/#{name}"
+    else
+      "#{var}/postgres"
```
This is repeated in a few places, might be nice extracting it.

cypher replied:
Adressed in dce8c390c3

MikeMcQuaid commented on Formula/postgresql@12.rb:
```diff
@@ -81,22 +81,87 @@ def post_install
+      caveats += if conflicts_with_postgresql_formula?
+        # Both PostgreSQL and PostgreSQL@12 are installed
+        <<~EOS
+          Previous versions of this formula used the same data directory as
```
```suggestion
          Previous versions of #{name} used the same data directory as
```

cypher replied:
Adressed in 2568ac4e20

MikeMcQuaid commented on Formula/postgresql@12.rb:
```diff
@@ -81,22 +81,87 @@ def post_install
+        # Both PostgreSQL and PostgreSQL@12 are installed
+        <<~EOS
+          Previous versions of this formula used the same data directory as
+          the regular PostgreSQL formula. This causes a conflict if you
```
```suggestion
          the postgresql formula. This causes a conflict if you
```

cypher replied:
Adressed in 2568ac4e20

MikeMcQuaid commented on Formula/postgresql@12.rb:
```diff
@@ -81,22 +81,87 @@ def post_install
 
+            mv -v "#{var}/postgres" "#{var}/#{name}"
+
+          (Make sure PostgreSQL is not running before executing this command)
```
```suggestion
          (Make sure PostgreSQL is stopped before executing this command)
```

cypher replied:
Adressed in 2568ac4e20

MikeMcQuaid commented on Formula/postgresql@12.rb:
```diff
@@ -81,22 +81,87 @@ def post_install
+    caveats += <<~EOS
       This formula has created a default database cluster with:
-        initdb --locale=C -E UTF-8 #{var}/postgres
+        initdb --locale=C -E UTF-8 #{var}/#{name}
```
Would be good to reuse one of the functions above.

cypher replied:
Addressed in 03dfafe1a4

MikeMcQuaid reviewed (commented):
Nice work, really like where this is going 👏🏻. Once we're agreed on the approach I think it might be good to use it universally.

cypher commented:
I've also expanded the `caveats` warning a bit that is printed if both postgres and postgresql@12 are installed, so it prints out the versioned data dir path. That way, someone who wants to migrate their 12 datadir to the versioned path will know where to put it.

MikeMcQuaid commented on Formula/postgresql@12.rb:
```diff
@@ -81,22 +81,100 @@ def post_install
+    end
+  end
+
+  def versioned_data_dir_exists?
```
```suggestion
  def versioned_pg_version_exists?
```

cypher replied:
Addressed in 66421b52a2

MikeMcQuaid commented on Formula/postgresql@12.rb:
```diff
@@ -81,22 +81,100 @@ def post_install
+  end
+
+  def versioned_postgresql_data_dir
+    "#{var}/#{name}"
```
```suggestion
    var/name
```

cypher replied:
Addressed in d2474d675c

MikeMcQuaid commented on Formula/postgresql@12.rb:
```diff
@@ -81,22 +81,100 @@ def post_install
+  end
+
+  def old_postgresql_data_dir
+    "#{var}/postgres"
```
```suggestion
    var/"postgres"
```

cypher replied:
Addressed in d2474d675c

MikeMcQuaid commented on Formula/postgresql@12.rb:
```diff
@@ -81,22 +81,100 @@ def post_install
+    end
+  end
+
+  def versioned_postgresql_data_dir
```
```suggestion
  def versioned_data_dir
```

cypher replied:
Addressed in 66421b52a2a35708f0a737e8360127ba32d414b4

MikeMcQuaid commented on Formula/postgresql@12.rb:
```diff
@@ -81,22 +81,100 @@ def post_install
+    "#{var}/#{name}"
+  end
+
+  def old_postgresql_data_dir
```
```suggestion
  def old_postgres_data_dir
```

cypher replied:
Addressed in 983c4475a63bd4e23dfabce8a2ae4cfc1e2d0734

MikeMcQuaid commented on Formula/postgresql@12.rb:
```diff
@@ -81,22 +81,100 @@ def post_install
+  # is version 12
+  def postgresql_log_path
+    if versioned_data_dir_exists?
+      "#{var}/log/#{name}"
```
```suggestion
      var/"log/#{name}"
```

cypher replied:
Addressed in d2474d675c

MikeMcQuaid commented on Formula/postgresql@12.rb:
```diff
@@ -81,22 +81,100 @@ def post_install
+    if versioned_data_dir_exists?
+      "#{var}/log/#{name}"
+    else
+      "#{var}/log/postgres"
```
```suggestion
      var/"log/postgres"
```

cypher replied:
Addressed in d2474d675c7de673046b9e01097094e3b60531a5

MikeMcQuaid commented on Formula/postgresql@12.rb:
```diff
@@ -81,22 +81,100 @@ def post_install
+  end
+
+  def versioned_data_dir_exists?
+    File.exist?("#{versioned_postgresql_data_dir}/PG_VERSION")
```
```suggestion
   ( versioned_postgresql_data_dir/"PG_VERSION").exist?
```

cypher replied:
Addressed in a608c8ba518efb1e88325c731817788bd29507b9

MikeMcQuaid commented on Formula/postgresql@12.rb:
```diff
@@ -81,22 +81,100 @@ def post_install
+    File.exist?("#{versioned_postgresql_data_dir}/PG_VERSION")
+  end
+
+  def conflicts_with_postgresql_formula?
```
I wouldn't use `conflicts_with` here as it's the name of a DSL not being used here.

cypher replied:
Addressed in f07001bcc5.

I've used `postgresql_formula_present?`, although it's not a great name, but I couldn't come up with anything better.

MikeMcQuaid commented on Formula/postgresql@12.rb:
```diff
@@ -81,22 +81,100 @@ def post_install
+  # Figure out what version of PostgreSQL the old data dir is
+  # using
+  def old_postgresql_datadir_version_12?
+    File.exist?("#{old_postgresql_data_dir}/PG_VERSION") &&
+      File.read("#{old_postgresql_data_dir}/PG_VERSION").chomp == "12"
```
```suggestion
    pg_version = old_postgresql_data_dir/"PG_VERSION"
    pg_version.exist? && pg_version.read.chomp == "12"
```

Might be nicer still to extract the `@12` from the name and use that.

cypher replied:
Addressed in 29942ac6d6.

I approached it a bit differently: This method now returns the version string from the `PG_VERSION` file (or nil if it doesn't exist), and then in the `caveats` method, I extract the formula's version number and compare the two.

MikeMcQuaid commented on Formula/postgresql@12.rb:
```diff
@@ -81,22 +81,100 @@ def post_install
+
+          In order to avoid this conflict, you should make sure that the
+          #{name} data directory is located at:
+
+            #{versioned_postgresql_data_dir}
```
```suggestion
            #{versioned_postgresql_data_dir}
```

cypher replied:
Addressed in 66421b52a2a35708f0a737e8360127ba32d414b4

MikeMcQuaid commented on Formula/postgresql@12.rb:
```diff
@@ -81,22 +81,100 @@ def post_install
+          try to use both at the same time.
 
+          You can migrate to a versioned data directory by running this command:
+
+            mv -v "#{old_postgresql_data_dir}" "#{versioned_postgresql_data_dir}"
```
```suggestion
            mv -v "#{old_postgresql_data_dir}" "#{versioned_postgresql_data_dir}"
```

cypher replied:
Addressed in 983c4475a63bd4e23dfabce8a2ae4cfc1e2d0734

MikeMcQuaid commented on Formula/postgresql@12.rb:
```diff
@@ -81,22 +81,100 @@ def post_install
+          the postgresql formula. This will cause a conflict if you
+          try to use both at the same time.
 
+          You can migrate to a versioned data directory by running this command:
```
```suggestion
          You can migrate to a versioned data directory by running:
```

cypher replied:
Addressed in 983c4475a63bd4e23dfabce8a2ae4cfc1e2d0734

MikeMcQuaid commented on Formula/postgresql@12.rb:
```diff
@@ -81,22 +81,100 @@ def post_install
+
+            mv -v "#{old_postgresql_data_dir}" "#{versioned_postgresql_data_dir}"
+
+          (Make sure PostgreSQL is stopped before executing this command)
+
```
```suggestion
          (Make sure PostgreSQL is stopped before executing this command)
```

cypher replied:
The reason I have the extra line in here is to create extra spacing with the later caveats string. Without it, it would "stick" together with the `This formula has created…` stuff, which is always printed. And I figured doing it this way would be more sensible than creating a two-element array, and joining that, or something.

MikeMcQuaid replied:
Gotcha, fine with me 👍🏻

MikeMcQuaid commented on Formula/postgresql@12.rb:
```diff
@@ -81,22 +81,100 @@ def post_install
+          #{name} data directory is located at:
+
+            #{versioned_postgresql_data_dir}
+
```
```suggestion
```

cypher replied:
Just FYI, same reason here as on line 163.

MikeMcQuaid reviewed (commented):
Nice work here!

cypher commented on Formula/postgresql@12.rb:
```diff
@@ -81,22 +81,101 @@ def post_install
-    unless File.exist? "#{var}/postgres/PG_VERSION"
-      system "#{bin}/initdb", "--locale=C", "-E", "UTF-8", "#{var}/postgres"
+    versioned_data_dir.mkpath
+    unless versioned_pg_version_exists?
+      system "#{bin}/initdb", "--locale=C", "-E", "UTF-8", versioned_data_dir
     end
```
The syntax CI job wants me to change these 3 lines into this:
```suggestion
    system "#{bin}/initdb", "--locale=C", "-E", "UTF-8", versioned_data_dir unless versioned_pg_version_exists?
```

But for me, that line is now more unreadable. So I've left it as-is.

MikeMcQuaid replied:
I'll need to be changed, sorry.

cypher replied:
Ok. Addressed in ae24ea6696.

MikeMcQuaid commented on Formula/postgresql@12.rb:
```diff
@@ -81,22 +81,101 @@ def post_install
+    caveats = ""
+
+    # Extract the version from the formula name
+    pg_formula_version = name.split("@", 2)[1]
```
```suggestion
    pg_formula_version = name.split("@", 2).last
```

MikeMcQuaid reviewed (commented):
This is looking good now! Could you now adjust the `PostgreSQL.rb` and other `PostgreSQL@*.rb` formulae to be consistent with this one? Can merge then, I think. Thanks!

cypher commented:
I've updated the other formulas. Since the others don't have the same problem as @12, their changes are simpler.

I've also incorporated the changes from https://github.com/Homebrew/homebrew-core/pull/65122, and I've updated the log path to include the `.log` at the end.

MikeMcQuaid commented:
> I've updated the other formulas. Since the others don't have the same problem as @12, their changes are simpler.

What is "the same problem"? I think they should all be modified to use version-specific datadirs, would you be able to do that?



> I've also incorporated the changes from #65122, and I've updated the log path to include the `.log` at the end.

Sorry, something I should have said: could you split this out to be one commit per formula? Thanks!

cypher commented:
> > I've updated the other formulas. Since the others don't have the same problem as @12, their changes are simpler.
> 
> What is "the same problem"? I think they should all be modified to use version-specific datadirs, would you be able to do that?

Sorry, I coulud have been more specific: The PostgreSQL@12 formula specifically did not use a verison-specifc datadir. All the other versioned formulas did. The (initial) changes in this PR are needed to deal with the case where the PostgreSQL@12 data dir might be conflicting with the regular PostgreSQL. All other versioned PostgreSQL formulas used a versioned data dir from the start, so they don't have this problem.

> > I've also incorporated the changes from #65122, and I've updated the log path to include the `.log` at the end.
> 
> Sorry, something I should have said: could you split this out to be one commit per formula? Thanks!

Will do. One commit per formula, or also one PR per formula?

MikeMcQuaid commented:
> Sorry, I should have been more specific: The PostgreSQL@12 formula specifically did not use a version-specifc datadir. All the other versioned formulas do. The (initial) changes in this PR are needed to deal with the case where the PostgreSQL@12 data dir might be conflicting with the regular PostgreSQL. All other versioned PostgreSQL formulas used a versioned data dir from the start, so they don't have this problem.

Gotcha 👍🏻 

> Will do. One commit per formula, or also one PR per formula?

One commit per formula, same PR is good 👍🏻.

If you're interested in taking on `mysql` after this: I'd love to see a new PR reviving https://github.com/Homebrew/homebrew-core/pull/60190 which takes the same approach that you have here!

cypher commented:
I've split up the changes for each formula into their own commits.

cypher commented:
> If you're interested in taking on `mysql` after this: I'd love to see a new PR reviving #60190 which takes the same approach that you have here!

Not much of a MySQL user, but I'll take a shot ;)

MikeMcQuaid commented:
Some failures but seem unrelated to these changes. Great work here @cyper! If you open the MySQL PR then please @mention me on it. Thanks again!

cypher commented:
Will do! Thanks for all the feedback!
